### DISCUSSION TITLE
Global hotkeys

### DISCUSSION THREAD

**Discussion by friday** [Ulauncher/Ulauncher#1347] [Troubleshooting]
Global hotkeys is a problem which cannot be solved universally for all platforms on Linux.

### Ulauncher version 6.x.x (currently in beta)
Ulauncher version 6.x.x supports setting global hotkeys for common Desktop Environments, by integrating with the Desktop Environments setting directly (supported for Gnome, KDE, XFCE, Budge, Pantheon aka Elementary OS). For other Desktop Environments that does not support this (like Cinnamon), or which we just haven't added an integration for, we will show the command you have to bind, but you have to figure out how to for your specific Desktop Environment on your own.

### Ulauncher version 5.x.x (and older)
Ulauncher version 5 implements hotkeys using [libkeybinder](https://github.com/kupferlauncher/keybinder). This only works on X11. For very valid security concerns, Wayland does not support applications listening to the keys the user types when it's not in focus. This however used to be the only way to support global hotkeys, and there are no replacements to this that works on all desktops.

libkeybinder can not override keyboard combinations set by the desktop environment. If you try to use for example alt+space and your desktop environment already has this mapped to something else, then you have to [change it in your desktop hotkey settings, or use another hotkey for Ulauncher](https://github.com/Ulauncher/Ulauncher/issues/222#issuecomment-402756636)

<p>For all these cases there is another way to add the keybindings yourself in the Desktop Environment or Window Manager settings:
<ol>
<li>Open Ulauncher Preferences and set hotkey to something you'll never use.</li>
<li>Bind the command <code>ulauncher-toggle</code> to the hotkey of your choice. Note that before Ulauncher version 6 is release you will also need the command <code>wmctrl</code> installed if you are getting focus issues with this command.</li>
</ol>
</p>
<details>
  <summary>Gnome instructions</summary>
<ul>
<li>Open <code>Settings > Keyboard</code> (may be named "Keyboard Shortcuts").</li>
<li>Scroll down to <code>Customize Shortcuts > Custom Shortcuts > +</code>.</li>
<li>In the "Command" field, enter <code>ulauncher-toggle</code>.</li>
<li>Set name and shortcut, then click <code>Add</code>.</li>
</ul>
</details>
</details>

### More complex workflows
Additional non-standard customizations like <a href="https://github.com/Ulauncher/Ulauncher/discussions/922">dual-role keys</a> or want to <a href="https://github.com/Ulauncher/Ulauncher/discussions/890">conditionally triggering the hotkey </a> would have to be set up with more complex custom workarounds or third party tools/services.

For a table detailing general Wayland compatibility, see https://github.com/Ulauncher/Ulauncher/discussions/1353

**Comment by NamelessCoder** [comment]
Small addendum about Gnome custom shortcut: for whatever reason, my Gnome (Ubuntu 24.04 + Ulauncher 6.0-beta15) was unable to set the global shortcut via the Ulauncher app (in fact the app reported a totally wrong shortcut when keys were pressed and ignored the space bar when used in combination with CTRL). I went to add the shortcut as a global shortcut in the Gnome settings (as described above for the 5.x solution) but this was not working at all. I tried with multiple different keystroke combinations, all with the same result.

Finally I decided to make a test with the full path to the `ulauncher-toggle` command and voila, it worked. So I'd add to the guide above:

* Some systems will not run the `ulauncher-toggle`  command unless it is entered with the full path.
* To get the full path of your command, type in terminal: `which unauncher-toggle` and copy this value to the shortcut setting's `command` field. For example, the value might be `/usr/bin/ulauncher-toggle`.

**Reply by friday** [reply]
I never had this issue, but I'm not on Ubuntu. So if someone else has to do this please comment to confirm.
